File: android/src/main/java/com/risemaxi/graft/classes/ManifestFile.java
package com.risemaxi.graft.classes;

import androidx.annotation.NonNull;
import org.json.JSONException;
import org.json.JSONObject;

public class ManifestFile {

    @NonNull
    private final String href;

    @NonNull
    private final String sha256;

    private final long size;

    public ManifestFile(@NonNull JSONObject json) throws JSONException {
        this.href = requireRelativePath(json.getString("href"));
        this.sha256 = json.getString("sha256");
        this.size = json.getLong("size");
        if (sha256.length() != 64) {
            throw new JSONException("sha256 must be a hex-encoded SHA-256 digest: " + sha256);
        }
        if (size < 0) {
            throw new JSONException("size must not be negative: " + size);
        }
    }

    @NonNull
    public String getHref() {
        return href;
    }

    @NonNull
    public String getSha256() {
        return sha256;
    }

    public long getSize() {
        return size;
    }

    @NonNull
    private static String requireRelativePath(@NonNull String href) throws JSONException {
        if (href.isEmpty() || href.startsWith("/") || href.contains("\\") || href.contains("//")) {
            throw new JSONException("href must be a relative path: " + href);
        }
        for (String segment : href.split("/")) {
            if (segment.equals(".") || segment.equals("..")) {
                throw new JSONException("href must not traverse directories: " + href);
            }
        }
        return href;
    }
}

File: ios/Sources/RisemaxiGraft/Classes/ManifestFile.swift
import Foundation

public struct ManifestFile: Decodable {
    let href: String
    let sha256: String
    let size: Int

    private enum CodingKeys: String, CodingKey {
        case href, sha256, size
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        href = try container.decode(String.self, forKey: .href)
        guard ManifestFile.isRelativePath(href) else {
            throw DecodingError.dataCorruptedError(forKey: .href, in: container, debugDescription: "href must be a relative path: \(href)")
        }
        sha256 = try container.decode(String.self, forKey: .sha256)
        guard sha256.count == 64 else {
            throw DecodingError.dataCorruptedError(forKey: .sha256, in: container, debugDescription: "sha256 must be a hex-encoded SHA-256 digest: \(sha256)")
        }
        size = try container.decode(Int.self, forKey: .size)
        guard size >= 0 else {
            throw DecodingError.dataCorruptedError(forKey: .size, in: container, debugDescription: "size must not be negative: \(size)")
        }
    }

    private static func isRelativePath(_ href: String) -> Bool {
        if href.isEmpty || href.hasPrefix("/") || href.contains("\\") || href.contains("//") {
            return false
        }
        return !href.split(separator: "/", omittingEmptySubsequences: false).contains { $0 == "." || $0 == ".." }
    }
}
